arangod/Agency/AsyncAgencyComm.cpp

```diff
@@ -0,0 +1,397 @@
+#include "Agency/AsyncAgencyComm.h"
+
+#include "ApplicationFeatures/ApplicationServer.h"
+#include "Futures/Utilities.h"
+#include "Basics/StaticStrings.h"
+#include "Scheduler/SchedulerFeature.h"
+#include "Logger/LogMacros.h"
+
+#include <chrono>
+
+#include <velocypack/Iterator.h>
+#include <velocypack/Buffer.h>
+#include <velocypack/velocypack-aliases.h>
+
+namespace {
+using namespace arangodb::fuerte;
+using namespace arangodb;
+
+using namespace std::chrono_literals;
+
+using clock = std::chrono::steady_clock;
+
+struct RequestMeta {
+  network::Timeout timeout;
+  fuerte::RestVerb method;
+  std::string url;
+  std::vector<std::string> clientIds;
+  network::Headers headers;
+  clock::time_point startTime;
+  unsigned tries;
+};
+
+bool agencyAsyncShouldCancel(RequestMeta &meta) {
+  if (meta.tries++ > 20) {
+    return true;
+  }
+
+  if (application_features::ApplicationServer::server().isStopping()) {
+    return true;
+  }
+
+  return false;
+}
+
+bool agencyAsyncShouldTimeout(RequestMeta const& meta) {
+  auto now = clock::now();
+  return (now > meta.startTime + meta.timeout);
+}
+
+auto agencyAsyncWaitTime(RequestMeta const& meta) {
+  if (meta.tries > 0) {
+    return 250ms;
+  }
+  return 0ms;
+}
```

Comment:
Perhaps use an exponential backoff? It's not that important with a fixed number of tries, but you could get the first retry faster, produce less load for multiple failures, and try for a longer time in one go. ;)